In [6]:
import importlib
import os
import pickle
import random

import numpy as np
import pandas as pd
import torch
import wandb
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from torch.utils.data import DataLoader
from torchvision.transforms import (
    Compose,
    RandomAffine,
    RandomHorizontalFlip,
    RandomResizedCrop,
    RandomRotation,
    RandomVerticalFlip,
)

from src import model, modules, utils

importlib.reload(model)
importlib.reload(modules)
importlib.reload(utils)

# pylint: disable = C0413
from src.model import MammoCNN
from src.utils import (
    BIRADS_MIN,
    LOSS_WEIGHTS,
    MammoCNNDataset,
    calibrate_loss_scales,
    effective_contributions,
    evaluate,
    get_coral_cutpoint_weights,
    get_findings_weights,
    get_lds_weights,
    train,
)

CLINICAL_FINDINGS = [
    "Mass",
    "Global Asymmetry",
    "Architectural Distortion",
    "Nipple Retraction",
    "Suspicious Calcification",
    "Focal Asymmetry",
    "Asymmetry",
    "Skin Thickening",
    "Suspicious Lymph Node",
    "Skin Retraction",
]

In [7]:
SEED: int = 42
BATCH_SIZE: int = 48
NUM_WORKERS: int = 8
DEVICE: torch.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
IMAGE_SIZE: list[int] = [1280, 1024]  # (H, W)


# Set seed to ensure consistency across runs
def seed_worker(worker_id):
    worker_seed = SEED + worker_id
    np.random.seed(worker_seed)
    random.seed(worker_seed)


g = torch.Generator()
g.manual_seed(SEED)
random.seed(SEED)
np.random.seed(SEED)

torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed(SEED)
    torch.cuda.manual_seed_all(SEED)

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

#### Split the dataset by using 600 patients for a hold out test set. The remaining 4,400 patients are randomly partitioned into a MAMMO CNN training set, a Classifier training set, a Triage training set, and a validation set of 60%, 15%, 15%, and 10%, respectively.

In [8]:
metadata_path = os.path.join("vindr-mammo", "breast_metadata.parquet")
metadata_df = pd.read_parquet(metadata_path)

# Create a patient level dataframe
patient_df = (
    metadata_df[["study_id", "patient_cancer", "Patient's Age"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

# Hold out test set 12%
train_val, test = train_test_split(
    patient_df,
    test_size=0.12,
    random_state=SEED,
    stratify=patient_df["patient_cancer"],
)

# Of the remaining 88%, split: 60% | 15% | 15% | 10%
# Sequential fractions:
#   val:             10/100 = 0.10  of remaining
#   triage_train:    15/90  ≈ 0.1667 of what's left after val
#   classifier_train:15/75  = 0.20  of what's left after val + triage
#   cnn_train:       60/60  = 1.0   (everything left)

# Split the remaining 4,400 patients
remaining_1, val = train_test_split(
    train_val,
    test_size=0.10,
    random_state=SEED,
    stratify=train_val["patient_cancer"],
)

remaining_2, triage_train = train_test_split(
    remaining_1,
    test_size=round(15 / 90, 6),  # ≈ 0.1667
    random_state=SEED,
    stratify=remaining_1["patient_cancer"],
)

cnn_train, classifier_train = train_test_split(
    remaining_2,
    test_size=round(15 / 75, 6),  # = 0.20
    random_state=SEED,
    stratify=remaining_2["patient_cancer"],
)

cnn_train.reset_index(drop=True, inplace=True)
classifier_train.reset_index(drop=True, inplace=True)
triage_train.reset_index(drop=True, inplace=True)
val.reset_index(drop=True, inplace=True)
test.reset_index(drop=True, inplace=True)

print(f"Number of patients in CNN training set: {len(cnn_train)}")
print(f"Number of patients in Classifier training set: {len(classifier_train)}")
print(f"Number of patients in Triage training set: {len(triage_train)}")
print(f"Number of patients in Validation set: {len(val)}")
print(f"Number of patients in Test set: {len(test)}")

Number of patients in CNN training set: 2639
Number of patients in Classifier training set: 660
Number of patients in Triage training set: 661
Number of patients in Validation set: 440
Number of patients in Test set: 600


#### Map the patient level splits back to the breast level dataframe

In [9]:
cnn_patients = cnn_train["study_id"].tolist()
classifier_patients = classifier_train["study_id"].tolist()
triage_patients = triage_train["study_id"].tolist()
val_patients = val["study_id"].tolist()
test_patients = test["study_id"].tolist()

# Map the patient level labels back to the original metadata
cnn_df = metadata_df[metadata_df["study_id"].isin(cnn_patients)].reset_index(drop=True)
classifier_df = metadata_df[
    metadata_df["study_id"].isin(classifier_patients)
].reset_index(drop=True)
triage_df = metadata_df[metadata_df["study_id"].isin(triage_patients)].reset_index(
    drop=True
)
val_df = metadata_df[metadata_df["study_id"].isin(val_patients)].reset_index(drop=True)
test_df = metadata_df[metadata_df["study_id"].isin(test_patients)].reset_index(
    drop=True
)

# Merge all training dataframes
train_df = pd.concat([cnn_df, classifier_df, triage_df], ignore_index=True)

# Standardize the age column
scaler = StandardScaler()
scaler.fit(train_df[["Patient's Age"]])
cnn_df["Patient's Age"] = scaler.transform(cnn_df[["Patient's Age"]])
classifier_df["Patient's Age"] = scaler.transform(classifier_df[["Patient's Age"]])
triage_df["Patient's Age"] = scaler.transform(triage_df[["Patient's Age"]])
val_df["Patient's Age"] = scaler.transform(val_df[["Patient's Age"]])
test_df["Patient's Age"] = scaler.transform(test_df[["Patient's Age"]])

# Save the age scaler
age_encoder_path = os.path.join("models", "age_encoder.pkl")

with open(age_encoder_path, "wb") as f:
    pickle.dump(scaler, f)

# Keep the std so age MAE can be reported in years, as the paper does (5.97 yrs)
AGE_STD: float = float(scaler.scale_[0])

cnn_df.rename(columns={"Patient's Age": "age"}, inplace=True)
classifier_df.rename(columns={"Patient's Age": "age"}, inplace=True)
triage_df.rename(columns={"Patient's Age": "age"}, inplace=True)
val_df.rename(columns={"Patient's Age": "age"}, inplace=True)
test_df.rename(columns={"Patient's Age": "age"}, inplace=True)

print(f"age std (years): {AGE_STD:.2f}")
print(f"BI-RADS range: {cnn_df['breast_birads'].min()}-{cnn_df['breast_birads'].max()}")
print(f"density range: {cnn_df['breast_density'].min()}-{cnn_df['breast_density'].max()}")

age std (years): 9.09
BI-RADS range: 1-5
density range: 1-4


In [ ]:
# 0-based ordinal levels for the weight helpers (the Dataset does this shift
# internally; here it is needed to count per-cutpoint positives)
cnn_suspicion_levels = cnn_df["breast_birads"].to_numpy() - BIRADS_MIN

# Per-cutpoint pos_weight for the ordinal (CORAL) suspicion head
suspicion_weights = get_coral_cutpoint_weights(cnn_suspicion_levels, n_cutpoints=4)

# Per-class weights for the multilabel findings head, clipped so the 17
# Skin Retraction positives cannot dominate the head
findings_matrix = np.stack(cnn_df["finding_vector"].to_numpy())
findings_weights = get_findings_weights(findings_matrix)

# LDS weights for density. kernel_size=3 because density has only 4 bins
density_lds = get_lds_weights(
    (cnn_df["breast_density"].to_numpy() - 1).tolist(),
    label_min=0,
    label_max=3,
    kernel_size=3,
    reweight="sqrt_inverse",
)
density_weights = torch.tensor(
    [density_lds[level] for level in range(4)], dtype=torch.float32
)

# MONAI applies alpha as: target * alpha + (1 - target) * (1 - alpha)
# so ALPHA is the weight on the POSITIVE (malignant) class and 1 - ALPHA the
# weight on the negative one. Two defensible readings:
#
#   ALPHA = POSITIVE_RATE -> positives 0.05, negatives 0.95  (pos:neg 0.05)
#       follows Lin et al., who found alpha < 0.5 best for the rare class at
#       gamma=2 because gamma already suppresses easy negatives hard.
#   ALPHA = 1 - POSITIVE_RATE  -> positives 0.95, negatives 0.05  (pos:neg 19.2)
#       inverse-frequency weighting, which is how the MAMMO paper describes
#       its alpha ("the inverse class frequency tuning parameter", Eq. 8).
#
# These differ by ~370x in the relative weight on malignancy, so it is worth
# running both and comparing diagnosis AP. Flip the line below to switch.
POSITIVE_RATE: float = float(cnn_df["diagnosis"].mean())
ALPHA: float = POSITIVE_RATE
# ALPHA: float = 1.0 - POSITIVE_RATE
GAMMA: float = 2.0

print("suspicion cutpoint pos_weights: ", [round(w, 2) for w in suspicion_weights.tolist()])
print("findings class weights: ", [round(w, 2) for w in findings_weights.tolist()])
print("density class weights: ", [round(w, 2) for w in density_weights.tolist()])
print(
    f"\nfocal alpha {ALPHA:.4f} -> malignant weight {ALPHA:.4f}, "
    f"benign weight {1 - ALPHA:.4f} (pos:neg = {ALPHA / (1 - ALPHA):.4f})"
)

loss_kwargs = {
    "loss_weights": LOSS_WEIGHTS,
    "suspicion_weights": suspicion_weights,
    "findings_weights": findings_weights,
    "density_weights": density_weights,
    "alpha": ALPHA,
    "gamma": GAMMA,
}

# Data transforms
ASPECT = IMAGE_SIZE[1] / IMAGE_SIZE[0]  # W / H
augmentations = Compose(
    [
        # Image flips
        RandomHorizontalFlip(p=0.5),
        RandomVerticalFlip(p=0.5),
        # Image rotations
        RandomRotation(degrees=20),
        # Image shear and shift of up to 20%
        RandomAffine(degrees=0, translate=(0.2, 0.2), shear=np.degrees(np.arctan(0.2))),
        # Image zoom of up to 20%
        RandomResizedCrop(size=IMAGE_SIZE, scale=(0.64, 1.0), ratio=(ASPECT, ASPECT)),
    ]
)

suspicion cutpoint pos_weights:  [2.0, 9.53, 19.22, 20.0]
findings class weights:  [0.27, 1.51, 0.8, 1.36, 0.4, 0.53, 0.88, 1.13, 1.11, 2.02]
density class weights:  [3.5, 1.2, 0.94, 1.08]

focal alpha 0.9505 -> malignant weight 0.9505, benign weight 0.0495 (pos:neg = 19.2222)


In [11]:
cnn_train_dataset = MammoCNNDataset(cnn_df, transform=augmentations, is_training=True)
val_dataset = MammoCNNDataset(val_df, is_training=False)
test_dataset = MammoCNNDataset(test_df, is_training=False)

cnn_train_loader = DataLoader(
    cnn_train_dataset,
    batch_size=BATCH_SIZE,
    shuffle=True,
    num_workers=NUM_WORKERS,
    pin_memory=True,
    worker_init_fn=seed_worker,
    generator=g,
    prefetch_factor=2,
    persistent_workers=True,
)
val_loader = DataLoader(
    val_dataset,
    batch_size=BATCH_SIZE,
    shuffle=False,
    num_workers=NUM_WORKERS,
    pin_memory=True,
    worker_init_fn=seed_worker,
    generator=g,
    prefetch_factor=2,
    persistent_workers=True
)
test_loader = DataLoader(
    test_dataset,
    batch_size=BATCH_SIZE,
    shuffle=False,
    num_workers=NUM_WORKERS,
    pin_memory=True,
    worker_init_fn=seed_worker,
    generator=g,
    prefetch_factor=2,
    persistent_workers=True
)

# Examine the train dataset
patient_batch = next(iter(cnn_train_loader))

print(f"Image Shape:     {patient_batch['mammogram'].shape}")
print(f"Diagnosis Shape: {patient_batch['diagnosis'].shape} {patient_batch['diagnosis'].dtype}")
print(f"Findings Shape:  {patient_batch['findings'].shape} {patient_batch['findings'].dtype}")
print(f"Suspicion Shape: {patient_batch['suspicion'].shape} {patient_batch['suspicion'].dtype}")
print(f"Density Shape:   {patient_batch['density'].shape} {patient_batch['density'].dtype}")
print(f"Age Shape:       {patient_batch['age'].shape} {patient_batch['age'].dtype}")

# Gradient accumulation exists to push this probability down: a batch with no
# positive contributes nothing to the diagnosis gradient. Below ~0.05 the
# effective batch is large enough and ACCUMULATION_STEPS can go to 1.
positive_rate = cnn_df["diagnosis"].mean()
for n in (BATCH_SIZE, BATCH_SIZE * 2):
    print(
        f"\npositive rate {100 * positive_rate:.2f}%  ->  "
        f"P(no positive in a batch of {n}) = {(1 - positive_rate) ** n:.4f}"
    )

Image Shape:     torch.Size([48, 3, 1280, 1024])
Diagnosis Shape: torch.Size([48]) torch.float32
Findings Shape:  torch.Size([48, 10]) torch.float32
Suspicion Shape: torch.Size([48]) torch.int64
Density Shape:   torch.Size([48]) torch.int64
Age Shape:       torch.Size([48]) torch.float32

positive rate 4.95%  ->  P(no positive in a batch of 48) = 0.0877

positive rate 4.95%  ->  P(no positive in a batch of 96) = 0.0077


#### Train the MAMMO CNN (Phase 1)

**Loss scale calibration.** The paper states that "loss weighting was adjusted according to the auxiliary output losses, such that the loss weight for diagnosis was greater than or equal to the sum of all other auxiliary output losses" (Appendix C-B). `LOSS_WEIGHTS` encodes that intent, but a coefficient only equals a share of the gradient if the raw losses share a scale — and they do not. Focal loss shrinks toward zero by design; CORAL sums weighted BCE over 4 cutpoints. Uncorrected, `diagnosis: 0.50` lands at ~4% of the signal while suspicion takes ~70%.

`calibrate_loss_scales()` measures each task's magnitude once on the freshly-initialised model and freezes the reciprocals. The coefficients then mean what they say: diagnosis really is 50% at step 0. The constants are deliberately **not** a running average — re-normalising every step would keep amplifying a task that has already converged, so no auxiliary would ever be allowed to finish.

Watch the `loss share` line: it should start at roughly 50/15/15/10/10 and then drift as tasks converge at different rates. Drift is expected and correct; a task getting easier legitimately gives up share.

**Gradient accumulation** gives an effective batch of `BATCH_SIZE * ACCUMULATION_STEPS`. At 4.95% positives a batch of 32 has a 19.6% chance of containing no positive at all; 64 drops that to 3.8%. On an A100 you can raise `BATCH_SIZE` instead and set `ACCUMULATION_STEPS = 1`.

In [12]:
EPOCHS: int = 100
LEARNING_RATE: float = 3e-4  # head LR
BACKBONE_LR_MULT: float = 0.1  # backbone trains at LEARNING_RATE * this
WARMUP_EPOCHS: float = 2.0
WEIGHT_DECAY: float = 1e-4
ACCUMULATION_STEPS: int = 2  # effective batch = BATCH_SIZE * ACCUMULATION_STEPS = 96

mammo_cnn = MammoCNN(
    n_findings=findings_matrix.shape[1],
    n_suspicion=5,
    n_density=4,
    pretrained=True,
)

# Deviation from the paper: the whole network trains from step 1 rather than
# cycling frozen/unfrozen stages. The paper's stated reason for cycling was
# preserving pretrained ImageNet features (Appendix C-B, after Shen 2017), not
# a memory limit — so the underlying risk still applies here. WARMUP_EPOCHS and
# BACKBONE_LR_MULT address it instead: nothing runs at full LR until the random
# heads have settled, and the backbone moves at a tenth of the head rate
# throughout. Call mammo_cnn.freeze_backbone() to reproduce the paper's regime.
print("full network:", mammo_cnn.get_parameter_counts())

# ── Calibrate loss scales so LOSS_WEIGHTS means gradient share ────────────────
# Run once, on the freshly-initialised model, then freeze.
loss_scales = calibrate_loss_scales(
    mammo_cnn, cnn_train_loader, DEVICE, loss_kwargs, n_batches=10
)
loss_kwargs["loss_scales"] = loss_scales

print("\nloss scales (1 / measured magnitude):")
for task, scale in loss_scales.items():
    print(f"  {task:>10}: {scale:9.4f}   (raw magnitude {1 / scale:.4f})")

# Sanity check: with the scales applied, shares should match LOSS_WEIGHTS
check = effective_contributions(
    {task: 1 / scale for task, scale in loss_scales.items()}, LOSS_WEIGHTS, loss_scales
)
print("\nshare at step 0 vs intended:")
for task in loss_scales:
    print(f"  {task:>10}: {100 * check[task]:5.1f}%  (target {100 * LOSS_WEIGHTS[task]:.0f}%)")

full network: {'total': 55897780, 'trainable': 55897780}

loss scales (1 / measured magnitude):
   diagnosis:   61.1633   (raw magnitude 0.0163)
    findings:    5.5075   (raw magnitude 0.1816)
   suspicion:    0.1951   (raw magnitude 5.1248)
     density:    0.7363   (raw magnitude 1.3582)
         age:    1.0953   (raw magnitude 0.9130)

share at step 0 vs intended:
   diagnosis:  50.0%  (target 50%)
    findings:  15.0%  (target 15%)
   suspicion:  15.0%  (target 15%)
     density:  10.0%  (target 10%)
         age:  10.0%  (target 10%)


In [13]:
# One-time setup, done in a terminal rather than here so the key never lands in
# a committed notebook:
#     pip install wandb
#     wandb login          # paste the key from https://wandb.ai/authorize
#
# On the SSH box prefer an env var so the run is non-interactive:
#     export WANDB_API_KEY=...
# If the node has no outbound network, set WANDB_MODE=offline and later run
# `wandb sync wandb/offline-run-*` from a machine that does.

ALPHA_STRATEGY = "positive_rate" if ALPHA < 0.5 else "one_minus_positive_rate"

# Name and config both derive from the actual values — a run labelled with a
# resolution or an alpha it was not trained at is worse than no label.
run = wandb.init(
    project="tulip-mammo-cnn",
    name=f"phase1-{IMAGE_SIZE[0]}x{IMAGE_SIZE[1]}-alpha-{ALPHA_STRATEGY}-lr-{LEARNING_RATE}",
    config={
        "phase": "cnn-phase1",
        "backbone": "inception_resnet_v2",
        "pretrained": True,
        "image_size": IMAGE_SIZE,
        "batch_size": BATCH_SIZE,
        "accumulation_steps": ACCUMULATION_STEPS,
        "effective_batch": BATCH_SIZE * ACCUMULATION_STEPS,
        "epochs": EPOCHS,
        "learning_rate": LEARNING_RATE,
        "backbone_lr_mult": BACKBONE_LR_MULT,
        "backbone_lr": LEARNING_RATE * BACKBONE_LR_MULT,
        "warmup_epochs": WARMUP_EPOCHS,
        "weight_decay": WEIGHT_DECAY,
        "freeze_schedule": "none (full network from step 1)",
        "seed": SEED,
        "loss_weights": LOSS_WEIGHTS,
        "loss_scales": loss_scales,
        "alpha": ALPHA,
        "alpha_strategy": ALPHA_STRATEGY,
        "malignant_weight": ALPHA,
        "benign_weight": 1 - ALPHA,
        "gamma": GAMMA,
        "suspicion_head": "coral-ordinal",
        "sampler": "none (class weighting in loss)",
        "suspicion_cutpoint_weights": suspicion_weights.tolist(),
        "findings_class_weights": findings_weights.tolist(),
        "density_class_weights": density_weights.tolist(),
        "n_train_images": len(cnn_df),
        "n_val_images": len(val_df),
        "positive_rate": POSITIVE_RATE,
    },
)

# Gradient and parameter histograms. Worth it here specifically to watch the
# CORAL cutpoint biases: if they stop being well separated the ordinal head has
# collapsed to a single threshold and suspicion is no longer ordinal.
wandb.watch(mammo_cnn, log="all", log_freq=100)

print(f"logging to {run.url}")

wandb: [wandb.login()] Loaded credentials for https://api.wandb.ai from WANDB_API_KEY.
wandb: Currently logged in as: moyinoluwa (moyinoluwa-university-of-lagos) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


logging to https://wandb.ai/moyinoluwa-university-of-lagos/tulip-mammo-cnn/runs/1obobtmr


In [ ]:
history = train(
    mammo_cnn,
    cnn_train_loader,
    val_loader,
    device=DEVICE,
    epochs=EPOCHS,
    learning_rate=LEARNING_RATE,
    weight_decay=WEIGHT_DECAY,
    accumulation_steps=ACCUMULATION_STEPS,
    warmup_epochs=WARMUP_EPOCHS,
    backbone_lr_mult=BACKBONE_LR_MULT,
    loss_kwargs=loss_kwargs,
    checkpoint_path=os.path.join("models", "mammo_cnn_phase1.pt"),
    monitor="ordinal_ap",
    monitor_mode="max",
    patience=7,
    lr_patience=3,
    lr_decay_factor=0.1,
    min_lr=1e-6,
    findings_names=CLINICAL_FINDINGS,
    age_std=AGE_STD,
    wandb_run=run,
)

# Store the best checkpoint as a versioned artifact so a run's weights stay
# tied to the config that produced them.
artifact = wandb.Artifact("mammo-cnn-phase1", type="model")
artifact.add_file(os.path.join("models", "mammo_cnn_phase1.pt"))
run.log_artifact(artifact)

epoch 1/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   1  train 0.8775  val 0.8577  lr[heads] 1.50e-04  lr[backbone] 1.50e-05  (warmup)
    diagnosis  AUC 0.4356  AP 0.0460   |  ordinal  AUC 0.5770  AP 0.0719
    findings   macroAP 0.0153  microAP 0.0447  macroAUC 0.4182
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0670  0.4898   105     1.12
      Suspicious Calcification      0.0338  0.4474    40     1.49
      Focal Asymmetry               0.0148  0.4626    28     0.93
      Architectural Distortion      0.0068  0.4430    13     0.92
      Asymmetry                     0.0027  0.2784     6     0.80
      Suspicious Lymph Node         0.0021  0.2185     5     0.74
      Nipple Retraction             0.0047  0.5347     4     2.07
      Skin Thickening               0.0017  0.3620     3     1.00
      Skin Retraction               0.0043  0.5276     3     2.51
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 0.84

epoch 2/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   2  train 0.8147  val 0.7938  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6327  AP 0.1027   |  ordinal  AUC 0.5614  AP 0.0645
    findings   macroAP 0.0135  microAP 0.0474  macroAUC 0.4576
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0593  0.4913   105     0.99
      Suspicious Calcification      0.0209  0.4673    40     0.92
      Focal Asymmetry               0.0175  0.5150    28     1.10
      Architectural Distortion      0.0094  0.5200    13     1.27
      Asymmetry                     0.0032  0.3508     6     0.93
      Suspicious Lymph Node         0.0046  0.4870     5     1.62
      Nipple Retraction             0.0028  0.4977     4     1.22
      Skin Thickening               0.0028  0.5331     3     1.62
      Skin Retraction               0.0015  0.2567     3     0.86
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 0.847  exact 0

epoch 3/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   3  train 0.7911  val 0.7823  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6032  AP 0.0960   |  ordinal  AUC 0.5977  AP 0.0772
    findings   macroAP 0.0193  microAP 0.0482  macroAUC 0.5380
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0593  0.4834   105     0.99
      Suspicious Calcification      0.0309  0.5988    40     1.36
      Focal Asymmetry               0.0129  0.4140    28     0.81
      Architectural Distortion      0.0064  0.4167    13     0.87
      Asymmetry                     0.0043  0.5066     6     1.26
      Suspicious Lymph Node         0.0441  0.6338     5    15.53
      Nipple Retraction             0.0030  0.4866     4     1.32
      Skin Thickening               0.0077  0.6348     3     4.52
      Skin Retraction               0.0048  0.6670     3     2.80
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.644  exact 0

epoch 4/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   4  train 0.7352  val 4.8148  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6013  AP 0.0856   |  ordinal  AUC 0.6013  AP 0.0857
    findings   macroAP 0.0145  microAP 0.0504  macroAUC 0.4553
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.0653  0.5004   105     1.09
      Suspicious Calcification      0.0249  0.4942    40     1.10
      Focal Asymmetry               0.0175  0.5082    28     1.10
      Architectural Distortion      0.0083  0.4200    13     1.12
      Asymmetry                     0.0034  0.4380     6     1.01
      Suspicious Lymph Node         0.0048  0.6145     5     1.68
      Nipple Retraction             0.0020  0.3021     4     0.86
      Skin Thickening               0.0017  0.3411     3     0.98
      Skin Retraction               0.0023  0.4792     3     1.34
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 3.528  exact 0

epoch 5/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   5  train 0.7019  val 0.7726  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6356  AP 0.2153   |  ordinal  AUC 0.6569  AP 0.2221
    findings   macroAP 0.0457  microAP 0.1165  macroAUC 0.6128
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1617  0.6485   105     2.71
      Suspicious Calcification      0.1640  0.6971    40     7.22
      Focal Asymmetry               0.0154  0.4507    28     0.97
      Architectural Distortion      0.0216  0.5186    13     2.93
      Asymmetry                     0.0080  0.5719     6     2.34
      Suspicious Lymph Node         0.0210  0.6217     5     7.39
      Nipple Retraction             0.0082  0.7150     4     3.60
      Skin Thickening               0.0065  0.6369     3     3.80
      Skin Retraction               0.0051  0.6549     3     2.98
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.743  exact 0

epoch 6/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   6  train 0.6891  val 1.6352  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6759  AP 0.1997   |  ordinal  AUC 0.6865  AP 0.2073
    findings   macroAP 0.0481  microAP 0.1145  macroAUC 0.6530
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1351  0.6165   105     2.26
      Suspicious Calcification      0.1995  0.7255    40     8.78
      Focal Asymmetry               0.0215  0.5456    28     1.35
      Architectural Distortion      0.0091  0.5013    13     1.23
      Asymmetry                     0.0384  0.7023     6    11.26
      Suspicious Lymph Node         0.0069  0.7328     5     2.43
      Nipple Retraction             0.0084  0.7985     4     3.70
      Skin Thickening               0.0106  0.7063     3     6.24
      Skin Retraction               0.0036  0.5483     3     2.12
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 3.522  exact 0

epoch 7/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   7  train 0.6689  val 0.8270  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7432  AP 0.2640   |  ordinal  AUC 0.7478  AP 0.2628
    findings   macroAP 0.0648  microAP 0.1449  macroAUC 0.7062
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1724  0.6765   105     2.89
      Suspicious Calcification      0.2468  0.7695    40    10.86
      Focal Asymmetry               0.0549  0.6383    28     3.45
      Architectural Distortion      0.0121  0.5209    13     1.64
      Asymmetry                     0.0055  0.5146     6     1.60
      Suspicious Lymph Node         0.0502  0.9287     5    17.68
      Nipple Retraction             0.0250  0.9260     4    10.99
      Skin Thickening               0.0081  0.7706     3     4.74
      Skin Retraction               0.0079  0.6105     3     4.66
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.074  exact 0

epoch 8/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   8  train 0.6549  val 0.7291  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6911  AP 0.2704   |  ordinal  AUC 0.6897  AP 0.2711
    findings   macroAP 0.0626  microAP 0.1619  macroAUC 0.6746
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2254  0.6596   105     3.78
      Suspicious Calcification      0.1956  0.7181    40     8.61
      Focal Asymmetry               0.0375  0.6777    28     2.35
      Architectural Distortion      0.0105  0.5764    13     1.42
      Asymmetry                     0.0047  0.4318     6     1.37
      Suspicious Lymph Node         0.0367  0.8141     5    12.93
      Nipple Retraction             0.0390  0.8868     4    17.14
      Skin Thickening               0.0062  0.6354     3     3.67
      Skin Retraction               0.0075  0.6714     3     4.39
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.710  exact 0

epoch 9/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch   9  train 0.6412  val 0.7704  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7480  AP 0.2388   |  ordinal  AUC 0.7582  AP 0.2420
    findings   macroAP 0.0744  microAP 0.1708  macroAUC 0.7104
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2649  0.7071   105     4.44
      Suspicious Calcification      0.1999  0.7412    40     8.79
      Focal Asymmetry               0.0523  0.6447    28     3.29
      Architectural Distortion      0.0117  0.5748    13     1.59
      Asymmetry                     0.0085  0.4625     6     2.50
      Suspicious Lymph Node         0.0897  0.9284     5    31.56
      Nipple Retraction             0.0280  0.8405     4    12.32
      Skin Thickening               0.0109  0.8919     3     6.40
      Skin Retraction               0.0034  0.6022     3     2.02
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.792  exact 0

epoch 10/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  10  train 0.6228  val 0.6707  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7352  AP 0.2860   |  ordinal  AUC 0.7515  AP 0.3014
    findings   macroAP 0.0652  microAP 0.1813  macroAUC 0.6910
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2656  0.6748   105     4.45
      Suspicious Calcification      0.1909  0.7433    40     8.40
      Focal Asymmetry               0.0448  0.6279    28     2.81
      Architectural Distortion      0.0102  0.5474    13     1.38
      Asymmetry                     0.0038  0.4635     6     1.13
      Suspicious Lymph Node         0.0365  0.8525     5    12.85
      Nipple Retraction             0.0223  0.9077     4     9.82
      Skin Thickening               0.0070  0.7904     3     4.13
      Skin Retraction               0.0061  0.6115     3     3.58
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.417  exact 0

epoch 11/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  11  train 0.6263  val 0.6883  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7329  AP 0.2880   |  ordinal  AUC 0.7564  AP 0.3084
    findings   macroAP 0.0584  microAP 0.1628  macroAUC 0.6457
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2507  0.6544   105     4.20
      Suspicious Calcification      0.1676  0.7295    40     7.38
      Focal Asymmetry               0.0281  0.6168    28     1.77
      Architectural Distortion      0.0094  0.4970    13     1.27
      Asymmetry                     0.0070  0.4414     6     2.04
      Suspicious Lymph Node         0.0280  0.6986     5     9.87
      Nipple Retraction             0.0228  0.7907     4    10.04
      Skin Thickening               0.0062  0.8025     3     3.61
      Skin Retraction               0.0060  0.5803     3     3.52
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.648  exact 0

epoch 12/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  12  train 0.6172  val 0.9553  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7323  AP 0.2044   |  ordinal  AUC 0.7410  AP 0.2159
    findings   macroAP 0.0491  microAP 0.1273  macroAUC 0.6419
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.1873  0.6641   105     3.14
      Suspicious Calcification      0.1741  0.7751    40     7.66
      Focal Asymmetry               0.0271  0.6361    28     1.71
      Architectural Distortion      0.0082  0.4868    13     1.11
      Asymmetry                     0.0036  0.3948     6     1.07
      Suspicious Lymph Node         0.0252  0.8091     5     8.85
      Nipple Retraction             0.0084  0.7489     4     3.69
      Skin Thickening               0.0042  0.7424     3     2.48
      Skin Retraction               0.0042  0.5196     3     2.45
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.298  exact 0

epoch 13/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  13  train 0.6104  val 0.7004  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7467  AP 0.3357   |  ordinal  AUC 0.7717  AP 0.3415
    findings   macroAP 0.0855  microAP 0.1964  macroAUC 0.6909
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2722  0.7029   105     4.56
      Suspicious Calcification      0.2983  0.7958    40    13.13
      Focal Asymmetry               0.0624  0.6941    28     3.92
      Architectural Distortion      0.0466  0.5334    13     6.31
      Asymmetry                     0.0078  0.5127     6     2.27
      Suspicious Lymph Node         0.0582  0.8897     5    20.47
      Nipple Retraction             0.0168  0.8377     4     7.38
      Skin Thickening               0.0040  0.6928     3     2.32
      Skin Retraction               0.0037  0.5593     3     2.14
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.551  exact 0

epoch 14/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  14  train 0.6079  val 0.7148  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.8045  AP 0.4661   |  ordinal  AUC 0.8174  AP 0.4776
    findings   macroAP 0.0996  microAP 0.2357  macroAUC 0.7436
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3134  0.7454   105     5.25
      Suspicious Calcification      0.3163  0.7770    40    13.92
      Focal Asymmetry               0.0530  0.6788    28     3.33
      Architectural Distortion      0.0083  0.5301    13     1.13
      Asymmetry                     0.0054  0.5792     6     1.59
      Suspicious Lymph Node         0.1380  0.9179     5    48.56
      Nipple Retraction             0.0382  0.8424     4    16.79
      Skin Thickening               0.0156  0.9051     3     9.17
      Skin Retraction               0.0082  0.7162     3     4.82
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.941  exact 0

epoch 15/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  15  train 0.5865  val 0.7212  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.6979  AP 0.3226   |  ordinal  AUC 0.7099  AP 0.3332
    findings   macroAP 0.0668  microAP 0.1748  macroAUC 0.6467
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2561  0.6628   105     4.29
      Suspicious Calcification      0.2134  0.7237    40     9.39
      Focal Asymmetry               0.0521  0.7036    28     3.28
      Architectural Distortion      0.0059  0.3805    13     0.80
      Asymmetry                     0.0122  0.5008     6     3.58
      Suspicious Lymph Node         0.0377  0.7658     5    13.26
      Nipple Retraction             0.0127  0.7218     4     5.57
      Skin Thickening               0.0084  0.8287     3     4.92
      Skin Retraction               0.0026  0.5327     3     1.54
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.633  exact 0

epoch 16/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  16  train 0.5923  val 0.6853  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7630  AP 0.3844   |  ordinal  AUC 0.7690  AP 0.3888
    findings   macroAP 0.0851  microAP 0.2223  macroAUC 0.7357
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3146  0.7284   105     5.27
      Suspicious Calcification      0.2589  0.7560    40    11.39
      Focal Asymmetry               0.0594  0.6401    28     3.73
      Architectural Distortion      0.0114  0.5229    13     1.54
      Asymmetry                     0.0152  0.7090     6     4.45
      Suspicious Lymph Node         0.0588  0.9462     5    20.68
      Nipple Retraction             0.0338  0.8411     4    14.86
      Skin Thickening               0.0108  0.8272     3     6.36
      Skin Retraction               0.0034  0.6507     3     1.98
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.665  exact 0

epoch 17/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  17  train 0.5852  val 0.9678  lr[heads] 3.00e-04  lr[backbone] 3.00e-05
    diagnosis  AUC 0.7440  AP 0.3961   |  ordinal  AUC 0.7482  AP 0.3997
    findings   macroAP 0.0826  microAP 0.1905  macroAUC 0.7238
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.2986  0.7359   105     5.01
      Suspicious Calcification      0.2833  0.7400    40    12.47
      Focal Asymmetry               0.0547  0.6067    28     3.44
      Architectural Distortion      0.0139  0.4999    13     1.89
      Asymmetry                     0.0102  0.7609     6     2.99
      Suspicious Lymph Node         0.0520  0.9560     5    18.32
      Nipple Retraction             0.0198  0.8134     4     8.71
      Skin Thickening               0.0073  0.8211     3     4.31
      Skin Retraction               0.0030  0.5802     3     1.77
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 2.327  exact 0

epoch 18/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  18  train 0.5793  val 0.8136  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7225  AP 0.3487   |  ordinal  AUC 0.7324  AP 0.3554
    findings   macroAP 0.0792  microAP 0.1903  macroAUC 0.7189
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3028  0.7253   105     5.08
      Suspicious Calcification      0.2259  0.7472    40     9.94
      Focal Asymmetry               0.0646  0.6469    28     4.06
      Architectural Distortion      0.0084  0.4876    13     1.14
      Asymmetry                     0.0070  0.6416     6     2.05
      Suspicious Lymph Node         0.0478  0.8266     5    16.83
      Nipple Retraction             0.0393  0.8825     4    17.27
      Skin Thickening               0.0123  0.9120     3     7.20
      Skin Retraction               0.0051  0.6003     3     2.97
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.718  exact 0

epoch 19/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  19  train 0.5361  val 0.7429  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7420  AP 0.3658   |  ordinal  AUC 0.7557  AP 0.3786
    findings   macroAP 0.0767  microAP 0.2098  macroAUC 0.7147
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3154  0.7328   105     5.29
      Suspicious Calcification      0.2292  0.7468    40    10.08
      Focal Asymmetry               0.0437  0.6440    28     2.75
      Architectural Distortion      0.0101  0.4750    13     1.37
      Asymmetry                     0.0053  0.5393     6     1.54
      Suspicious Lymph Node         0.0391  0.8824     5    13.76
      Nipple Retraction             0.0288  0.8629     4    12.68
      Skin Thickening               0.0138  0.9152     3     8.11
      Skin Retraction               0.0046  0.6338     3     2.71
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.645  exact 0

epoch 20/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  20  train 0.5340  val 0.8226  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7262  AP 0.3594   |  ordinal  AUC 0.7397  AP 0.3700
    findings   macroAP 0.0757  microAP 0.2037  macroAUC 0.6831
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3140  0.7224   105     5.26
      Suspicious Calcification      0.2334  0.7508    40    10.27
      Focal Asymmetry               0.0512  0.6454    28     3.22
      Architectural Distortion      0.0095  0.4611    13     1.28
      Asymmetry                     0.0037  0.4883     6     1.08
      Suspicious Lymph Node         0.0345  0.8410     5    12.16
      Nipple Retraction             0.0252  0.8401     4    11.09
      Skin Thickening               0.0066  0.8374     3     3.90
      Skin Retraction               0.0035  0.5612     3     2.03
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.752  exact 0

epoch 21/100:   0%|          | 0/220 [00:00<?, ?it/s]

epoch  21  train 0.5213  val 0.8236  lr[heads] 3.00e-05  lr[backbone] 3.00e-06
    diagnosis  AUC 0.7355  AP 0.3597   |  ordinal  AUC 0.7469  AP 0.3769
    findings   macroAP 0.0753  microAP 0.2009  macroAUC 0.7085
      per class (by support)            AP     AUC     n  AP/prev
      Mass                          0.3046  0.7224   105     5.11
      Suspicious Calcification      0.2183  0.7577    40     9.60
      Focal Asymmetry               0.0510  0.6528    28     3.20
      Architectural Distortion      0.0097  0.4608    13     1.32
      Asymmetry                     0.0043  0.5614     6     1.27
      Suspicious Lymph Node         0.0499  0.8993     5    17.56
      Nipple Retraction             0.0269  0.8373     4    11.82
      Skin Thickening               0.0102  0.8886     3     5.96
      Skin Retraction               0.0031  0.5959     3     1.79
      Global Asymmetry                  --      --     0       --   (no positives in split)
    suspicion  MAE 1.706  exact 0

epoch,▁▁▂▂▂▃▃▃▄▄▅▅▅▆▆▆▇▇▇██
train/loss,█▇▆▅▅▄▄▄▃▃▃▃▃▃▂▂▂▂▁▁▁
train/loss_share/age,▄█▆▂▁▁▁▂▂▃▃▃▄▃▅▄▄▄▇▇█
train/loss_share/density,█▅▅▅▃▂▂▁▂▁▁▁▁▁▂▁▁▂▄▄▄
train/loss_share/diagnosis,▁▆▆▇▇███▇▇▇▇▇▇▆▇▇▇▅▅▅
train/loss_share/findings,█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
train/loss_share/suspicion,▁▃▃▄▄▃▃▃▅▄▅▅▄▅▆▅▅▅███
train/lr/backbone,▄████████████████▁▁▁▁
train/lr/heads,▄████████████████▁▁▁▁
train/raw_loss/age,██▇▄▃▃▃▂▂▂▂▂▂▂▂▂▂▁▁▁▁
+54,...


In [ ]:
# Per-class findings report
final = evaluate(
    mammo_cnn, val_loader, DEVICE, loss_kwargs, CLINICAL_FINDINGS, AGE_STD
)

report = pd.DataFrame(
    {
        "support": final["findings_support"],
        "AUROC": final["findings_auc_per_class"],
        "AP": final["findings_ap_per_class"],
    }
).sort_values("support", ascending=False)
report["prevalence %"] = 100 * report["support"] / len(val_df)

print("── findings, per class (validation) ──")
print(report.to_string(float_format=lambda v: f"{v:.4f}"))
print(f"\nmacro AP {final['findings_macro_ap']:.4f}   micro AP {final['findings_micro_ap']:.4f}")

print("\n── suspicion, per BI-RADS level (paper Table I format) ──")
for level, auc in final["suspicion_auc_per_level"].items():
    print(f"  {level}: AUROC {auc:.4f}")

print("\n── comparison with the paper (Tommy dataset, Table I) ──")
print(f"  diagnosis AUROC   ours {final['ordinal_auc']:.3f}   paper 0.787")
print(f"  diagnosis AUPRC   ours {final['ordinal_ap']:.3f}   paper 0.394")
print(f"  age MAE (years)   ours {final['age_mae_years']:.2f}    paper 5.97")
print("  (the paper used TTA over 100 samples; these are single-pass)")

run.finish()

── findings, per class (validation) ──
                          support  AUROC     AP  prevalence %
Mass                          105 0.7454 0.3134        5.9659
Suspicious Calcification       40 0.7770 0.3163        2.2727
Focal Asymmetry                28 0.6788 0.0530        1.5909
Architectural Distortion       13 0.5301 0.0083        0.7386
Asymmetry                       6 0.5792 0.0054        0.3409
Suspicious Lymph Node           5 0.9179 0.1380        0.2841
Nipple Retraction               4 0.8424 0.0382        0.2273
Skin Thickening                 3 0.9051 0.0156        0.1705
Skin Retraction                 3 0.7162 0.0082        0.1705
Global Asymmetry                0    NaN    NaN        0.0000

macro AP 0.0996   micro AP 0.2357

── suspicion, per BI-RADS level (paper Table I format) ──
  birads_1: AUROC 0.6227
  birads_2: AUROC 0.4559
  birads_3: AUROC 0.4593
  birads_4: AUROC 0.4330
  birads_5: AUROC 0.9596

── comparison with the paper (Tommy dataset, Table I) ──
  